package/utils/nvram/files/nvram-bcm53xx.init

```diff
@@ -39,11 +39,44 @@ set_wireless_led_behaviour() {
 	esac
 }
 
+set_bcm43602_variables() {
+	# set variables needed for Broadcom BCM43602
+
+	case $(board_name) in
+		asus,rt-ac3200)
+			COMMIT=1
+
+			# radio 0 (5 GHz)
+			nvram set devpath0=pcie/1/3
+			nvram set 0:devpath0=sb/1/
+			nvram set 0:devid=0x43bc
+			nvram set 0:sromrev=11
+			nvram set 0:boardflags=0x30040000
+
+			# radio 1 (2.4 GHz)
+			nvram set devpath1=pcie/1/4
+			nvram set 1:devpath1=sb/1/
+			nvram set 1:devid=0x43bb
+			nvram set 1:boardrev=0x1421
+			nvram set 1:sromrev=11
+			nvram set 1:boardflags=0x20001000
+
+			# radio 2 (5 GHz)
+			nvram set devpath2=pcie/2/1
+			nvram set 2:devpath2=sb/1/
+			nvram set 2:devid=0x43bc
+			nvram set 2:sromrev=11
+			nvram set 2:boardflags=0x30040000
+			;;
```

Comment:
Will the wifi driver read this from the system nvram?
This looks like some standard configuration is there no wifi configuration in the nvram at all? where does the vendor firmware get these values from?

Reply:
> Will the wifi driver read this from the system nvram?

Yes, brcmfmac reads these at load to set up the wireless PHYs.

> This looks like some standard configuration is there no wifi configuration in the nvram at all?

Tom discovered that when NVRAM is reset, these variables disappear. So we need to set it on OpenWrt.

> where does the vendor firmware get these values from?

From some proprietary Broadcom documentation, I'd guess.

Reply:
You should not reset the nvram, this contains essential configurations for the system. Sometimes the nvram gets restored from some other storage, but this is often not fully implemented. Maybe you should check before setting these variables if they are set and do not overwrite them in that case. I think it is sufficient to check if the boardflags are set. If you can find these settings in the nvram, do not set all the values. 

Do you have a dump of the nvram before and after resetting it?

Reply:
Hauke, I think you misunderstood what I meant. I don't mean emptying out all variables on NVRAM when I say resetting NVRAM. The CFE bootloader has an option called "reset NVRAM". That reset option supplies the default system configuration. The nvram-bcm53xx.init file is for supplying the missing configuration.

With that said, Tom has properly tested and confirmed that these variables that I set are absolutely required for setting up the wireless PHYs. So I believe that this part of the patch series is proper.

Reply:
At least in the past when I worked in brcm the nvram reset in CFE did not work very well. It often reset it to some boardcom defaults and not to something the vendor needs.

Reply:
The brcmfmac and other Broadcom drivers are using the sprom which is stored in the nvram to configure the wifi card. 
I think we should not write to nvram if it is not necessary. If someone deletes the NVRAM then they should manually restore it again. I do not want to write some values into it just to fix a device someone broke.

The NVRAM on broadcom is comparable to the calibration data on other devices.

if we detect that the nvram is broken for sure we can maybe fix it, but we should not unconditionally write into it.